locator/abstract_replication_strategy.hh

```diff
@@ -126,10 +147,64 @@ public:
     future<dht::token_range_vector> get_pending_address_ranges(const token_metadata_ptr tmptr, std::unordered_set<token> pending_tokens, inet_address pending_address, locator::endpoint_dc_rack dr) const;
 };
 
+using replication_strategy_ptr = seastar::shared_ptr<const abstract_replication_strategy>;
+using mutable_replication_strategy_ptr = seastar::shared_ptr<abstract_replication_strategy>;
+
+/// \brief Represents effective replication (assignment of replicas to keys).
+///
+/// It's a result of application of a given replication strategy
+/// over a particular token metadata version, for a given table.
+/// Can be shared by multiple tables if they have the same replication.
+///
+/// Immutable, users can assume that it doesn't change.
+///
+/// Holding to this object keeps the associated token_metadata_ptr alive,
+/// keeping the token metadata version alive and seen as in use.
+class effective_replication_map {
+protected:
+    replication_strategy_ptr _rs;
+    token_metadata_ptr _tmptr;
+    size_t _replication_factor;
+public:
+    effective_replication_map(replication_strategy_ptr, token_metadata_ptr, size_t replication_factor);
+    virtual ~effective_replication_map() = default;
+
+    const abstract_replication_strategy& get_replication_strategy() const noexcept { return *_rs; }
+    const token_metadata& get_token_metadata() const noexcept { return *_tmptr; }
+    const token_metadata_ptr& get_token_metadata_ptr() const noexcept { return _tmptr; }
+    const topology& get_topology() const noexcept { return _tmptr->get_topology(); }
+    size_t get_replication_factor() const { return _replication_factor; }
+
+    virtual inet_address_vector_replica_set get_natural_endpoints(const token& search_token) const = 0;
+    virtual inet_address_vector_replica_set get_natural_endpoints_without_node_being_replaced(const token& search_token) const = 0;
+    virtual inet_address_vector_topology_change get_pending_endpoints(const token& search_token, const sstring& ks_name) const = 0;
+    virtual std::unique_ptr<token_range_splitter> make_splitter() const = 0;
+};
+
+using effective_replication_map_ptr = seastar::shared_ptr<const effective_replication_map>;
+using mutable_effective_replication_map_ptr = seastar::shared_ptr<effective_replication_map>;
+
+/// Replication strategies which support per-table replication extend this trait.
+///
+/// It will be accessed only if the replication strategy actually works in per-table mode,
+/// that is after mark_as_per_table() is called, and as a result
+/// abstract_replication_strategy::is_per_table() returns true.
+class per_table_replication_strategy {
```

Comment:
It's to avoid the "diamond problem" in case replication strategy implements multiple traits.

Reply:
> And why not make `is_per_table()` and `maybe_as_per_table()` virtual functions, which would be overriden in `per_table_replication_strategy`? (then we could remove `_per_table` bool) Is it due to performance considerations?

This is to avoid the cost of virtual calls.